src/api/build_type.rs

```diff
@@ -0,0 +1,67 @@
+use crate::Engine;
+
+/// Trait to build a custom type for use with the [`Engine`].
+/// i.e. register the type, getters, setters, methods, etc...
+///
+/// # Example
+///
+/// ```
+/// use rhai::{Engine, RhaiCustomType};
+///
+/// #[derive(Debug, Clone, Eq, PartialEq)]
+/// struct TestStruct {
+///     field: i64
+/// }
+///
+/// impl TestStruct {
+///     fn new() -> Self {
+///         Self { field: 1 }
+///     }
+///     fn update(&mut self, offset: i64) {
+///         self.field += offset;
+///     }
+/// }
+///
+/// impl RhaiCustomType for TestStruct {
+///     fn build(engine: &mut Engine) {
+///         engine
+///             .register_type::<Self>()
+///             .register_fn("new_ts", Self::new)
+///             .register_fn("update", Self::update);
+///     }
+/// }
+///
+/// # fn main() -> Result<(), Box<rhai::EvalAltResult>> {
+///
+/// let mut engine = Engine::new();
+///
+/// // Register API for the custom type.
+/// engine.build_type::<TestStruct>();
+///
+/// # #[cfg(not(feature = "no_object"))]
+/// assert_eq!(
+///     engine.eval::<TestStruct>("let x = new_ts(); x.update(41); x")?,
+///     TestStruct { field: 42 }
+/// );
+/// # Ok(())
+/// # }
+/// ```
+pub trait RhaiCustomType {
+    /// Builds the custom type for use with the [`Engine`].
+    /// i.e. register the type, getters, setters, methods, etc...
+    fn build(engine: &mut Engine);
+}
```

Comment:
Let's spend a little time thinking about what other functions may be needed.  Once this goes out, we can no longer easily change this trait.

Reply:
Yes, I think it's wise.

Also I thought of not expose directly `Engine` to avoid potiential side effects and using a wrapper that only expose methods related to building type.

Here's just a prototype I have in mind
```rust
struct TypeBuilder<'a, T> {
    engine: &'a mut Engine,
    name: Option<&'static str>,
    _marker: PhantomData<T>,
};

impl<'a, T> TypeBuilder<'a> {
    fn with_name(&mut self, name: &'static str) -> &mut self {
        self.name = Some(name);
        self
    }

    fn add_method<N, F>(&mut self, name: N, method: F) -> &mut self 
    where /* omited */
    {
        self.engine.register_fn(name, method);
        self
    }

    // other method to add getters, setters, etc.

    fn build(self) {
        /* empty */
    }
}

impl<'a, T> Drop for TypeBuilder<'a, T> {
    fn drop(&mut self) {
        if let Some(name) = self.name {
            self.engine.register_type_with_name::<T>(name);
        } else {
            self.engine.register_type::<T>();
        }
    }
}
```

Reply:
I have the feeling that `engine.build_type::<Type>()` looks the easiest...

Reply:
I implement the prototype on another branch.
Here's what I have in mind: https://github.com/tguichaoua/rhai/blob/4432d5063a44c082d1f20c4c4889865b927ed7d9/tests/build_type.rs

Reply:
This actually looks quite interesting!  I would suggest go with this one instead!

Also, why not mark it (and all the related types) as `volatile` (usually I use a specially-phrased `#[deprecated]` attribute  to do that) so that users know that this API might change.

Reply:
Why not use a feature like `type_builder` and state in the documentation that this feature is unstable and may change in the future ?
Once stabilized, we can remove the `#[cfg(feature = "type_builder)]` and in the documentation say that the `type_builder` feature does nothing and will be removed in the next major update.

Reply:
Well, we would actually want to encourage the user to use it without worrying about changing code later on.

If we force the user to pull in a feature, then once that feature goes away, it will be a breaking change.

If we simply mark it as volatile, however, the user can simply use it, and as long as we don't change the API, there is nothing he needs to do when we  make it non-volatile - the warning just disappears.

Reply:
Ok